include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp

```diff
@@ -0,0 +1,462 @@
+                m_polygon_projections[oss.str()] = polygon_projection{projected_polygon, base, v1,
+                    v2, normal};
+            }
+
+            // Check if the projected polygon is valid
+            if (!resolve_dynamic::is_valid<polygon_2d>::apply(projected_polygon, visitor, strategy))
+            {
+                return intersection_type::invalid;
+            }
+        }
+
+        // Check if the segment intersects the plane of the polygon
+        CT dot_product = dot(normal, base);
+        CT sign1 = dot(normal, p1) - dot_product;
+        CT sign2 = dot(normal, p2) - dot_product;
+
+        if ((sign1 > tolerance && sign2 > tolerance) || (sign1 < -tolerance && sign2 < -tolerance))
+        {
+            return intersection_type::none; // No intersection, this is valid
+        }
+
+        // Check if the segment is on the same plane as the polygon
+        if (std::abs(sign1) < tolerance && std::abs(sign2) < tolerance)
+        {
+            auto projected_p1 = project_point_to_plane_2d(v1, v2, p1);
+            auto projected_p2 = project_point_to_plane_2d(v1, v2, p2);
+
+            linestring_2d projected_segment = {projected_p1, projected_p2};
+
+            // Check if the projected segment intersects the polygon
+            std::tuple
+            <
+                boost::geometry::model::multi_point<point_2d>,
+                boost::geometry::model::multi_linestring<linestring_2d>> intersections;
+            boost::geometry::intersection(projected_segment, projected_polygon, intersections);
+
+            auto& points = std::get<0>(intersections);
+            auto& linestrings = std::get<1>(intersections);
+
+            if (boost::geometry::is_empty(points) && boost::geometry::is_empty(linestrings))
+            {
+                return intersection_type::none; // No intersection, this is valid
+            }
+            if (!points.empty())
+            {
+                visitor.template apply<failure_invalid_intersection>();
+                return intersection_type::invalid;
+            }
+            if (linestrings.size() == 1)
+            {
+                if (boost::geometry::equals(linestrings.front(), projected_segment))
+                {
+                    if (!has_correct_intersection_orientation(projected_polygon, projected_p1,
+                                                              projected_p2, visitor))
+                    {
+                        return intersection_type::invalid;
+                    }
+                    return intersection_type::valid; // The only case of valid non empty intersection
+                }
+            }
+            visitor.template apply<failure_invalid_intersection>();
+            return intersection_type::invalid;
+        }
+
+        // Compute the intersection point
+        vec3 seg_dir = p2 - p1;
+        CT denom = dot(normal, seg_dir);
+        assert(denom != 0); // Ensure the segment is not parallel to the plane
```

Comment:
Is this necessarily a bug where the program should be halted in a build with asserts? Seems like something that could also happen with e.g. invalid coordinate inputs (e.g. denormal small coords so that product is zero, since normal is not normalised afaics) that should not be a halt but the check returning an is_valid failure.